### PR TITLE
[RLlib] Move Models + ActorCriticEncoder + Tests

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

In order to make https://github.com/ray-project/ray/pull/32069/files#diff-c5bcc55e59b3d9007b1231ad8a624f9def304f2d200cda2eef6088f4150f53b5 easier to review, this PR isolates the changes to models and introduces tests.

This PR makes it such that the base models and torch models, including their configs are well documented and well tested and moved into the critical path of the rllib/core/ folder.

### CODE REVIEW

File: rllib/core/models/tf/primitives.py
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tf/primitives.py:82]: this is not a limitation at the moment.

File: rllib/core/models/tf/primitives.py
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tf/primitives.py:27]: The PR is mostly about TorchModels and BaseModels, as the descriptions says.
This TF Model is not tested, but I wanted to take some of the torch changes already and get this a little up to speed. These changes make it so that it at least conforms to what we find in Torch.

File: rllib/algorithms/ppo/tests/test_ppo_rl_module.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ppo/tests/test_ppo_rl_module.py:73]: We "traditionally" write these in lower case in model_config and later convert them to something like ReLU in framework-specific code. This is now happening here as well.

File: rllib/algorithms/ppo/tf/ppo_tf_rl_module.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:58]: The complete setup is moved into a PPORLModule base class in the upcoming pr!

File: rllib/algorithms/ppo/tf/ppo_tf_rl_module.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:96]: The changes to the forward functions are not algorithmic, but separate encoder, policy- and value-head code a little nicer. I had to rewrite this anyway because the model interfaces changed.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/base.py:87]: This is obviously the same example as the above.
This makes sense because the classes are very tightly coupled (so ought to be used together).
Still, both should contain an example in their docstring.

File: rllib/core/models/configs.py
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/configs.py:13]: I'm using this decorator as long as I'm developing these models to clearly filter the stuff that does not go with TF.
We can potentially remove it after reachon parity between TF/Torch.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/base.py:122]: This is in the last file of this diff.
Please let me know if you can think of a more elegant solution.
Requirements: Make sure that a model without any requirements to inputs/outputs does not error out when we check specs on the forward method.

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha This PR contains a subset of the changes requested by you on the other PR. I've answered them there and made the changes for every comment that has a 👍 and marked ones where further discussion is needed with 👎

File: rllib/models/specs/specs_base.py
Comment (kouroshHakha) [inline rllib/models/specs/specs_base.py:31]: empty list or SpecDict should do the same thing. This is not needed imo

**Review by kouroshHakha — APPROVED** [review]
Nice PR man! A few comments and questions before merging. Also for these type of PRs that are smaller pieces of a larger PR, make sure the reviewer is aware what is gonna come next and how things are broken down to begin with.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:126]: The thing is that SpecDict barfs if it does not get a mapping to check.
My first reaction was that always requiring a mapping as output for all models is not a restriction we want to impose on people.
Especially folks who just want to write some custom models would be weirded out by the spec checking errors this produces if they specifically have no specs defined.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:29]: Changed to:

> A ModelConfig is usually built by RLModules after getting it from a Catalog object.
    It is therefore a means of configuration for RLModules. However, ModelConfigs are
    not tied to usage with Catalog or RLModules.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:70]: Your are right, CNN imposes other requirements here.
I've tackled this in the next PR:
https://github.com/ray-project/ray/pull/32515

These configs are for the high-level models that we use as building blocks for RLModules.
I therefore believe that we don't need a CNNModels that output 3D-tensors but rather vectors.
We only need CNNs in two places: As encoders, as tokenizers. Both should output vectors.
Lmk if you believe we should further discuss this.

File: rllib/core/models/configs.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py:111]: Thanks for the heads up! I'll make sure to do this.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:223]: Wrote this while RLModules inputs specs depended on this and needed to .`update()` with more specs (which does not work for a simple list). Changed! Thanks

File: rllib/core/models/configs.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py:128]: Thanks, changed!

File: rllib/core/models/configs.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py:153]: Sorry, just renamed this to "IdentityEncoderConfig".
This can be, in theory, used if people just want completely distinct actor and critic networks.
That is, networks that have a different architecture.
I'll add this to a docstring.

File: rllib/core/models/torch/tests/test_mlp.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/tests/test_mlp.py:25]: Yeah, just to signal that this is a list of hidden_layer_dims. 
[hidden_layer_dims_1, hidden_layer_dims_2, ...]
where hidden_layer_dims_1 is a list of hidden dims.
I saw this somewhere else in RLlib and thought it made sense. I admit that it's not super clear, lmk if you can think of something cooler.

File: rllib/core/models/torch/tests/test_encoder.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/tests/test_encoder.py:1]: Yeah, I've split this up in my upcoming PR (drafy hidden_layer_dims). This test only takes 4.195 seconds now on my MBP, but it's also simply more comfortable to have different files for these tests. It's going to be 4-6 files!

File: rllib/core/models/base.py
Comment (kouroshHakha) [reply rllib/core/models/base.py:126]: how does convert_to_canonical work when an empty list is given to it? We want the output of that, right?

Comment (kouroshHakha) [reply rllib/core/models/base.py:126]: here: https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/models/specs/checker.py?L82&subtree=true
the output would be `SpecDict({})`

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:126]: Huh. I was not aware of that tool (`convert_to_canonical_format`), sorry.
Let me read through that first and see if I can apply it here.

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:126]:
```
from ray.rllib.core.models.base import Model, ModelConfig
from ray.rllib.models.specs.checker import check_input_specs
from ray.rllib.models.specs.checker import check_output_specs
from dataclasses import dataclass

class MyModel(Model):
    def __init__(self, config):
        super().__init__(config)
        self.my_param = config.my_param * 2

    @check_input_specs("input_spec")
    @check_output_specs("output_spec")
    def _forward(self, input_dict):
        return input_dict["obs"] * self.my_param


@dataclass
class MyModelConfig(ModelConfig):
    my_param: int = 42

    def build(self, framework: str):
        if framework == "bork":
            return MyModel(self)


config = MyModelConfig(my_param=3)
model = config.build(framework="bork")
print(model._forward({"obs": 1}))
```

This code results in the following error:

```
File "/Users/artur/code/ray/python/ray/rllib/models/specs/checker.py", line 152, in _validate
    raise ValueError(f"{tag} must be a Mapping, got {type(data).__name__}")
ValueError: output must be a Mapping, got int
```

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:126]: .... with:
```
Model.input_spec = convert_to_canonical_format([])
Model.output_spec = convert_to_canonical_format([])
```

Comment (kouroshHakha) [reply rllib/core/models/base.py:126]: looking ...

Comment (kouroshHakha) [reply rllib/core/models/base.py:126]: After offline discussion, both checker decorators need to only check specs if the specs are not None. So if we make the base class by default return None, then the specs won't get checked and user won't have to worry about it.

Comment (kouroshHakha) [reply rllib/core/models/base.py:126]: This will get rid of needing a DummySpec()

File: rllib/core/models/base.py
Comment (kouroshHakha) [reply rllib/core/models/base.py:29]: not tied to usage with Catalog or RLModules --> not restricted to be used only with Catalog or RLModules

File: rllib/core/models/base.py
Comment (kouroshHakha) [reply rllib/core/models/base.py:70]: Got it, I think I understood the requirement. Anything beyond this use-case demands a custom RLModule. Let's revisit this again in the next PR.

File: rllib/core/models/configs.py
Comment (kouroshHakha) [reply rllib/core/models/configs.py:153]: Are we gonna use this in RLllib code base? If it's just left for examples, let's remove it in this PR? I think shared = True / False already does that, don't it?

File: rllib/core/models/torch/tests/test_mlp.py
Comment (kouroshHakha) [reply rllib/core/models/torch/tests/test_mlp.py:25]: when that happens the variable name should be list_of_hidden_layer_dims instead of hidden_layer_dimss :)

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/tests/test_mlp.py:25]: Or... a list_of_list_of_hidden_layer_dim.

File: rllib/core/models/torch/mlp.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/mlp.py:12]: So for MLP, the primitive is used in both, MLPEncoder and MLPModel.
For MLPModel and MLPEncoder I think this makes sense especially because it strictly unifies their configs and the logic that builds the layers. MLPModel and MLPEncoder only wrap the primitive model. So if we were to get rid of the primitive MLP, Encoder would have to wrap MLPModel in order to avoid code duplication. Would that be better in your opinion? 

I think having three MLP classes makes sense but admit that it could be clearer in terms of documentation.

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/mlp.py:12]: Going with `TorchMLPHead` here now, because I believe head is more easily and universally understood.

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha I've made the following main changes after your last review:

1. ActorCriticEncoder is now a base class for TorchActorCriticEncoder and TfActorCriticEncoder (they share 99% of logic)
2. I implemented https://github.com/ray-project/ray/pull/32453#discussion_r1106255892 (with a super slight twist to make the specs accessible via Model.input_spec and Model.output_spec)
3. MLPModel is now MLPHead and the tests look accordingly

The rest where nits I believe.

File: rllib/core/models/base.py
Comment (kouroshHakha) [inline rllib/core/models/base.py:100]: nit (do only if you gotta make changes because of some other reasons): decorate with overrideToMakeCustomization flag

File: rllib/core/models/base.py
Comment (kouroshHakha) [inline rllib/core/models/base.py:112]: nit: same re decoration

File: rllib/core/models/torch/base.py
Comment (kouroshHakha) [inline rllib/core/models/torch/base.py:71]: This is weird. sub-classes should always call super().__init__() at the very beginning.

File: rllib/core/models/base.py
Comment (kouroshHakha) [inline rllib/core/models/base.py:305]: nit: to make it easier to read this can just be a normal dict. doesn't have to be NestedDict

File: rllib/core/models/base.py
Comment (kouroshHakha) [inline rllib/core/models/base.py:325]: same here.

File: rllib/core/models/base.py
Comment (kouroshHakha) [inline rllib/core/models/base.py:356]: Just FYI. You can return dicts and they will get converted to NestedDict due to spec checking decoration

File: rllib/core/models/configs.py
Comment (kouroshHakha) [inline rllib/core/models/configs.py:53]: I think torch uses weird mix (Camel-case maybe): ReLU

File: rllib/core/models/configs.py
Comment (kouroshHakha) [inline rllib/core/models/configs.py:183]: I still think we should get rid of this Identity encoder in this PR, it won't be used in any of the upcoming PRs, right?

Comment (kouroshHakha) [reply rllib/core/models/configs.py:183]: Would catalog use this in any way?

File: rllib/core/models/torch/encoder.py
Comment (kouroshHakha) [inline rllib/core/models/torch/encoder.py:187]: For example this should call 

```
TorchModel().__init__()
ActorCriticEncoder().__init__()
```

Comment (kouroshHakha) [reply rllib/core/models/torch/encoder.py:187]: what's wrong with this?

**Review by kouroshHakha — COMMENTED** [review]
One important note about multiple inheritance and the super() call. the rest are nits.

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:100]: Damn I was looking for that one and could not find it.

Comment (kouroshHakha) [reply rllib/core/models/base.py:100]: use source graph man :) https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/utils/annotations.py?L110:5&subtree=true

File: rllib/core/models/torch/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/base.py:71]: Yes. But here are the issues:

- We need to instantiate the wrapped encoder(s) in order to get their specs.
- So we need to do something like `self.encoder = blah` and AFTER that something like `self.input_spec = self.encoder.input_spec()`. 
- The problem is that we need to call nn.Module.__init__() before `self.encoder = blah` because torch enforces it. 
- But that also means that we can not call in in TorchModel anymore, because that would delete `self.encoder`.
- If we were to put TorchModel.__init__() BEFORE all of this, it would call Model.__init__() which in turn already calls `self.get_input_spec()` (which does not work because it needs `self.encoder` in the case of ActorCrtiticEncoder)

File: rllib/core/models/base.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/base.py:356]: Thanks! Yeah, I just did this because it would otherwise violate the return type.

File: rllib/core/models/configs.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py:53]: Yep, sorry!

File: rllib/core/models/configs.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py:183]: Nope, I'm getting rid of it.

File: rllib/core/models/torch/encoder.py
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/encoder.py:187]: See https://github.com/ray-project/ray/pull/32453#discussion_r1106566241

File: rllib/core/models/torch/base.py
Comment (kouroshHakha) [reply rllib/core/models/torch/base.py:71]: That's why we have to call get_input_specs() only in `__post_init__()` :) Only after instantiating the whole module you compute the specs that depend on other specs.

Comment (kouroshHakha) [reply rllib/core/models/torch/base.py:71]: I have gone through all of these thought process once with RLModules, man :). There is a reason for that `__postinit__()` plugin that I added to the constructor of the base class.

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/base.py:71]: 😄

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/base.py]: @kouroshHakha Ok, changed. Python is beatiful.

Comment (kouroshHakha) [reply rllib/core/models/torch/base.py]: Nice 👍

File: rllib/BUILD
Comment (ArturNiederfahrenhorst) [inline rllib/BUILD:1865]: @kouroshHakha Added this since your last review. I'll merge these tests once we have feature parity but will only create torch tests for the next models for now and leave this as is.

File: rllib/tuned_examples/ppo/pendulum-ppo-with-rl-module.yaml
Comment (ArturNiederfahrenhorst) [inline rllib/tuned_examples/ppo/pendulum-ppo-with-rl-module.yaml:23]: When you first introduced this test, the tf models used relus regardless of configuration.
Together with this fixed seed this made for an apparently advantagious situation.

The standard for activation function is actually tanh all over RLlib afaics (I wonder why that is, sounds pretty slow).
In order to make this test pass, I tried to reproduce the exact model that we used when you introduced the test, hence this config value.

**Review by kouroshHakha — APPROVED** [review]
LGTM for merge

**Comment by kouroshHakha** [comment]
@ArturNiederfahrenhorst Why did everyone got tagged?

File: rllib/tuned_examples/ppo/pendulum-ppo-with-rl-module.yaml
Comment (kouroshHakha) [reply rllib/tuned_examples/ppo/pendulum-ppo-with-rl-module.yaml:23]: I looked at this, a little bit. I think the tf thing is slow not because of activation but because of sth else. I think we should disable this test for now and re-enable when eager-tracing works on the rllib sampling side. So for now I would just comment the tf part of the test out and move on.

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha Happended during a revert of master. 🤷

**Pull request merged**
